File: 00-leetcode/src/链表/_206_反转链表.java
package 链表;
/**
 * 反转一个单链表。
 * 
 * 示例:
 *		输入: 1->2->3->4->5->NULL
 *		输出: 5->4->3->2->1->NULL
 *
 * 使用递归或迭代两种方式解决
 * 
 * 迭代：正想思维，即一种从头到尾的思考模式：for、while循环等
 * 递归：逆向思维，即一种从尾向前的思考模式：函数自己调自己
 * 
 * note：整体思路：
 * 		1.新建一个空链表 hader->null
 * 		2.遍历原链表，将遍历到的元素插入到新链表头部
 * 
 * 参考思路：https://blog.csdn.net/FX677588/article/details/72357389
 * 
 * https://leetcode-cn.com/problems/reverse-linked-list/
 * @author nius
 *
 */
public class _206_反转链表 {

	//Definition for singly-linked list.
	public static class ListNode {
		int val;
		ListNode next;
		ListNode(int x) { val = x; }
	}
	
	/// --------------------- 迭代模式 ---------------------
	// head->1->2->3->4->5->NULL
	// head->5->4->3->2->1->NULL
	public static ListNode reverseList(ListNode head) {
		// 空链表 or 只有一个节点
        if (head == null || head.next == null) return head;
		
        // 迭代
        // 1.新建一个空链表
        ListNode newHead = null; 
        
        // 2.遍历原链表，将遍历到的元素插入新链表头部
        /**
         * 0.初始状态
         *  head->1->2->3->4->5->null
         *  newHead->null
         * 1.第一次循环
         * head == 1 不为空进入循环
         * temp == 2
         * head.next -> 2 ==> head.next -> null === (1 - > null) 
         * newHead = head ==> newHead -> head == newHead -> 1
         * head = temp ==> head -> 2
         * ----> newHead->1->null
         * 
         * 2.第二次循环
         * head == 2 不为空进入循环
         * temp == 3
         * head.next -> 3 ==> head.next -> 1 === (2 -> 1) 
         * newHead = head ==> newHead -> head == newHead -> 2
         * head = temp ==> head -> 3
         * ----> newHead->2->1->null
         * 
         * ...
         */
        while (head != null) {
			ListNode temp = head.next; // 先记录head的下一个节点
			head.next = newHead;
			newHead = head;
			head = temp;
		}
		return newHead;
    }
	
	/// --------------------- 递归模式 ---------------------
	// head->1->2->3->4->5->NULL
	// head->5->4->3->2->1->NULL
	public static ListNode re_reverseList(ListNode head) {
		if (head == null || head.next == null) return head;
		ListNode newHead = re_reverseList(head.next); //一直循环到链尾 
		
		// note：此处需要等到re_reverseList子层的栈结束才会返回
		head.next.next = head; //翻转链表的指向
		head.next = null; //记得赋值NULL，防止链表错乱		
		return newHead; //新链表头永远指向的是原链表的链尾
	}
	
	/**
	{
	head->1->2->3->null
		{
		head->2->3->null
		
			{
			head->3->null
			// head.next == null成立
			// 直接返回head
			// 此时结构如此：head->3->null
			// 1.------->代码来到这里return head，此栈中后边的三句代码不会执行
			
			head.next.next = head; 
			head.next = null; 		
			return newHead; 
			} // 此栈结束，返回到上一层栈
			
		// 2.------->代码来到这里
		// 返回的head赋值给newHead
		// 此时newHead->3->null，head->2->3->NULL，执行后边代码
		
		// head=2,head.next=3,head.next.next=3.next
		head.next.next = head;  ==> 3->2
		head.next = null; 	==> 3->null	
		return newHead; 
		
		// 3.------->代码来到这里
		// newHead->3->null
		}
		
	head.next.next = head; 
	head.next = null; 		
	return newHead; 
	}
	 */
	
	
	
	public static void printLinkList(ListNode head) {
		ListNode tempHead = head;
		StringBuilder string = new StringBuilder();
		string.append("head->");
		
		while (tempHead != null) {
			string.append(tempHead.val).append("->");
			tempHead = tempHead.next;
		}
		string.append("null");
		System.out.println(string);
	}
	
	public static void main(String[] args) {
		
		// 创建链表 1->2->3->4->5->null
		ListNode node1  = new ListNode(1);
		ListNode node2 = new ListNode(2);
		node1.next = node2;
		ListNode node3 = new ListNode(3);
		node2.next = node3;
		ListNode node4 = new ListNode(4);
		node3.next = node4;
		ListNode node5 = new ListNode(5);
		node4.next = node5;
		ListNode head = node1;

		// 翻转前
		printLinkList(head);

		// 翻转后	
		ListNode newHead = reverseList(head);
		printLinkList(newHead);
		
		// 再递归翻转
		ListNode newHead2 = re_reverseList(newHead);
		printLinkList(newHead2);
	}

}

File: LeetCode/链表/_206_单链表反转/_206_反转链表.m
//
//  _206_反转链表.m
//  链表
//
//  Created by 曾维俊 on 2019/5/6.
//  Copyright © 2019年 Nius. All rights reserved.
//

#import "_206_反转链表.h"
#import "ListNode.h"

/**
 单链表反转
 https://leetcode-cn.com/problems/reverse-linked-list/
 
 输入: 1->2->3->4->5->NULL
 输出: 5->4->3->2->1->NULL
 
 递归+迭代实现
 
 迭代：while循环实现
 
 递归：函数嵌套调用
 过程分析:head->1->2->3->NULL
 思路为：
    1.当前函数栈保存head后，让head往后移动，重复此操作，知道head移动到最后一个
    2.由于函数栈是嵌套关系，最后的head先出栈，依次往前，将出栈后的head返回给上一个
    3.每出一次栈，会对外一层栈中的头两个节点执行反转操作
        3.1 newHead->5->4->null
        3.2 head->3->4->5->null
        3.3 head->3->4->3 让头节点的下一个节点指向头结点，导致newHead跟着变化newHead->5->4->3，此时实际上是:
            newHead->5->4->3
            head->3->4->3（3和4连个节点其实是一个环，下一步需要断掉该环）
        3.4 head->3->null（断掉环），同时newHead会变为newHead->5->4->3->null
 
    4. 步骤3会不断执行，直到最外层栈结束，返回newHead即为翻转后的结果
 
    【该思路巧妙运用栈先进后出原则，让head向后移动开栈的之前记住当期head，实现】
 
 
 
   1.开栈帧1
    {
        head->1->2->3->NULL（参数）
        if (不满足) ...往下执行

            2.开栈帧2
            {
                head->2->3->NULL（参数）
                if (不满足) ...往下执行
 
                    3.开栈帧3
                     {
                        head->3->NULL（参数）
                        if (满足 head.next == NULL) 直接return 返回 head->3->NULL
                     } 4.栈帧3销毁（只有一个元素，没做翻转）
 
 
                5.来到这里
                5.1 此时newHead就是栈帧3的返回值：newHead->3->NULL
                5.2 此时的head->2->3->NULL
 
                head.next.next = head; ==> 2->3->2 (此时结构 head->2->3->2,==> newHead->3->2，即有两个指针同时指向3)
                head.next = nil; ==>断掉head中2指向3的指针（断掉循环结构），此时head->2->NULL,newHead->3->2->NULL
                return 函数返回：newHead->3->2->NULL
            } 6.栈帧2销毁（只有2个元素，翻转，元素为链表最后两个元素，即实现最后两个元素的翻转）
 
        6.来到这里
        6.1 此时newHead就是栈帧2的返回值：newHead->3->2->NULL
        6.2 head->1->2->3->NULL
 
        head.next.next = head;  ==> 1->2->1 (此时结构 head->1->2->1,==> newHead->3->2->1，即有两个指针同时指向3)
        head.next = nil; ==>断掉head中1指向2的指针（断掉循环结构），此时head->1->NULL,newHead->3->2->1->NULL
        return 函数返回：newHead->3->2->NULL（只有2个元素，翻转，次元素为链表最后两个元素，即实现最后两个元素的翻转）
    }7.栈帧1销毁（实现前两个元素翻转，即实现整个链表的翻转）
 */

// 递归
ListNode *re_reverseList(ListNode *head) {
    if (head == nil || head.next == nil) return head;
    
    // 使head向后移动，并开辟新的栈帧（新栈帧获取到的head是向后移动后的值）
    ListNode *newHead = re_reverseList(head.next);
    
    // 使用当前栈帧中的head
    head.next.next = head;
    head.next = nil;
    return newHead;
}

// 迭代
ListNode *reverseList(ListNode *head) {
    if (head == nil || head.next == nil) return head;
    
    ListNode *newHead = nil; //空链表
    while (head != nil) {
        // 1.记住下一个head下一个，方便head向后移动
        ListNode *temp = head.next;
        
        // 2.将head节点插入newHead头部
        head.next = newHead;
        newHead = head;
        
        // 3.head向后移动
        head = temp;
    }
    
    return newHead;
}

@implementation _206_____

+ (void)main {
    
    NSMutableArray *numbers = @[].mutableCopy;
    for (NSInteger i = 0; i < 10; i++) {
        [numbers addObject:@(i)];
    }
    ListNode *head = createList(numbers);
    
    printList(head, @"递归翻转前:head");
    // 递归实现
    ListNode *head1 = re_reverseList(head);
    printList(head1, @"递归翻转后:head");
    
    // 迭代实现
    ListNode *head2 = reverseList(head1);
    printList(head2, @"迭代再翻转:head");
}

@end
